Enable rubric editing in manual grading assessment question page

Victorsss-Orz opened the pull request:
Closes #7562, part of #12260. The ability to modify the rubric in manual grading assessment question page is helpful not only in AI grading mode but also in human manual grading mode, as we don't need to have a student submission before being able to create or update the rubric. 

None-editing mode by default on page load
<img width="1883" height="421" alt="image" src="https://github.com/user-attachments/assets/bb483467-df1f-414b-ae1a-15a32576a472" />
Enabling editing. The panel for general settings is also collapsible, with an arrow at the right to indicate
<img width="1881" height="533" alt="image" src="https://github.com/user-attachments/assets/71d6614b-c0f7-4a79-8e02-1a79288fc083" />
Support rubric creation
<img width="1886" height="334" alt="image" src="https://github.com/user-attachments/assets/6d54d820-e695-4e10-b5f2-6fdaa19b021c" />

jonatanschroeder commented:
I believe you can tag #7562 for closing in this PR.

Victorsss-Orz commented on apps/prairielearn/assets/scripts/instructorAssessmentManualGradingRubricEditingPreact.tsx:
```diff
@@ -0,0 +1,28 @@
+  // assert(rubricTable.dataset.assessmentQuestion);
+  if (!rubricTable.dataset.csrfToken || !rubricTable.dataset.assessmentQuestion) {
+    return;
+  }
```
I'm using this in place of an assertion. For some reason I can't get the import to work for assert.

reteps replied:
This is a client side script, it can't assert. You should throw an error instead.

reteps commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessmentQuestion/assessmentQuestion.html.tsx:
```diff
@@ -150,67 +150,34 @@ export function AssessmentQuestion({
         : ''}
 
+      <div class="mb-3">
+        <div
```
`hydrateHtml`

Victorsss-Orz commented on apps/prairielearn/assets/scripts/instructorAssessmentManualGradingRubricSettings.tsx:
```diff
@@ -0,0 +1,24 @@
+  const __csrf_token = rubricSettings.dataset.csrfToken;
+  const assessment_question = JSON.parse(rubricSettings.dataset.assessmentQuestion);
+  const rubric_data = JSON.parse(rubricSettings.dataset.rubricData ?? 'null');
+  const aiGradingStats = JSON.parse(rubricSettings.dataset.aiGradingStats ?? 'null');
```
I'm using hydrateHtml directly in the page so don't need this extra script. The variables are already well-defined in that page.

reteps commented:
Here are some initial things I noticed:

<img width="2448" height="806" alt="CleanShot 2025-08-25 at 09 34 23@2x" src="https://github.com/user-attachments/assets/337ee435-0c07-4b12-9ab1-04432189ddc8" />

- The "Add Item" button isn't in line with the other buttons for rubric editing
- If I delete an item, and cancel the deletion, the page doesn't reload, so it incorrectly shows that the item is still deleted till I reload.
- It is not clear to me what the "Disable Rubric" button does. Should that have a "?" icon next to it explaining it's purpose

- The buttons could benefit from icons (e.g. a plus for add, a gear for 'rubric settings', a pencil for edit)

<img width="2472" height="1134" alt="CleanShot 2025-08-25 at 09 38 34@2x" src="https://github.com/user-attachments/assets/25a507e8-c38d-49e6-914e-2370a1710944" />

- The radio options should be laid out vertically, not horizontally
- The minimum/maximum rubric score/extra credit should be laid out vertically
- Clicking 'Edit rubric' could open this expander by default. It wasn't clear to me that that it contained useful settings/info at first, it looks like a section header.
  - To fix this, can we give the section a header, like "Rubric"
     - And add another section header, like "Rubric points"

reteps commented:
> Should I do "Show" and "Hide" separately based on the whether the content is collapsed, or the same text is fine?

I think the same text is fine. Can you add some spacing -- "Hide / Show", and move the button to the right edge of the header?

<img width="1300" height="112" alt="CleanShot 2025-09-10 at 09 59 48@2x" src="https://github.com/user-attachments/assets/3f53a0b4-3eae-4da0-a04b-5cf2625160e3" />

Victorsss-Orz commented:
Also addressing the rest of the things in the review: 

- The detailed explanation and grader note don’t show correctly on first load (e.g. I added a field, not editable).
  - Fixed
- The AI Agreement column should not be shown in rubric settings.
  - This was agreed upon in earlier dev meetings that went over the design. We specifically want "AI agreement" column in AI grading mode to help instructor make modifications to the rubric items when trying AI grading. In non-AI grading mode, it will still be the "In use" as the current rubric setting modal. 
- The disable rubric should be renamed to ‘delete rubric’, and should delete all items. It should have red text, not blue
  - Text changed. The behavior of the rubric deletion is reused from the current version. The rubric itself is soft deleted, and the items are kept but not reusable since the rubric is not available. 
- The drag handler for the up/down should change the cursor to the ‘grabbed’ state
  - It's sometimes working and sometimes not. I'll look at how to make it consistent. 
- The rubric items themselves should be within the rubric settings dropdown, and the dropdown should use a hide/show button instead of just a basic accordion
  - Fixed

Victorsss-Orz commented:
<img width="1671" height="226" alt="image" src="https://github.com/user-attachments/assets/6e106b38-7a3d-4f75-91ac-242adb89d03e" />
This is actually existing js in the page already so I'm using it to deal with the text change and animation.

Victorsss-Orz commented on apps/prairielearn/src/components/RubricSettings.tsx:
```diff
@@ -0,0 +1,624 @@
+      body: JSON.stringify(payload),
+    });
+    if (!res.ok) {
+      const data = (await res.json().catch(() => ({ err: `Error: ${res.statusText}` }))) ?? {};
```
Addressed in 1ef0287b6

Victorsss-Orz commented on apps/prairielearn/src/components/RubricSettings.tsx:
```diff
@@ -0,0 +1,624 @@
+      <div id="rubric-setting" class="js-collapsible-card-body show p-2">
+        {/* Settings */}
+        <div>
+          {!!assessmentQuestion.max_auto_points && (
```
Addressed in 1ef0287b6

reteps replied:
cc @nwalters512: I'm thinking that we want to get rid of using "!!value" if possible.

Victorsss-Orz commented on apps/prairielearn/src/components/RubricSettings.tsx:
```diff
@@ -0,0 +1,624 @@
+          role="button"
+          tabIndex={editMode ? 0 : -1}
+          aria-disabled={!editMode}
+          class={`btn btn-sm btn-ghost ${!editMode ? 'disabled' : ''}`}
```
Addressed in 1ef0287b6

Victorsss-Orz commented on apps/prairielearn/src/components/RubricSettings.tsx:
```diff
@@ -0,0 +1,624 @@
+        <td class="align-middle">
+          {submissionCount === 0 ? (
+            <span>&mdash;</span>
+          ) : item.disagreement_count === undefined || item.disagreement_count === null ? (
```
Addressed in 1ef0287b6

Victorsss-Orz commented on apps/prairielearn/src/components/RubricSettings.tsx:
```diff
@@ -0,0 +1,624 @@
+        </td>
+      ) : (
+        <td class="text-nowrap align-middle">
+          {item.num_submissions === undefined
```
Addressed in 1ef0287b6

reteps replied:
The line this talks about was not.

Victorsss-Orz replied:
Uh oh. f9f3d41d1

Victorsss-Orz commented on apps/prairielearn/src/components/RubricSettings.tsx:
```diff
@@ -0,0 +1,624 @@
+                type="button"
+                class="btn btn-secondary me-2"
+                onClick={() => {
+                  setRubricItems(rubricItemDataMerged);
```
Updated in f9f3d41d1

reteps commented on apps/prairielearn/src/components/RubricSettings.tsx:
```diff
@@ -0,0 +1,624 @@
+  const moveUp = (idx: number) => {
+    if (idx === 0) return;
+    setRubricItems((prev) => {
+      const next = prev.slice();
```
To be clear, it shouldn't cause any issues to do that.

nwalters512 replied:
I don't think it's worth "enforcing", but I do think that the callback form is very useful at times. I don't want to mandate either form and I think devs should just do what they deem appropriate in a given scenario.
